### PR TITLE
Add support for writing geometries to WKB

### PR DESCRIPTION
**Pull request by meastp**
This patchset adds support for writing to the WKB format from geometries. There are a couple of fixes to support geometry adaptions there as well, and I had to modify the geometry_type to support different dimensions and formats.

This implementation is generic enough to be able to support CIRCULARSTRING, implemented on top of this implentation locally, so it should be sufficiently user-extendable for some use cases.

Current limitations: 
* endian conversion is not supported (yet)
* there might be some use of C++11 in there, which I will remove when the direction of this PR
* no support for multi-geometries (yet)
* there might be traces of CIRCULARSTRING left.

I am prepared to do more work to get this patch into Boost.Geometry proper, as it is a hassle to maintain this out-of-tree. :)

### CODE REVIEW

**Comment by meastp** [comment]
Is this merged yet? If not, is there anything I can do to accelerate the process? :)

**Comment by awulkiew** [comment]
AFAIU the only reason why this wasn't merged yet is that @mloskot wanted to do it. I think there are no objections. @mloskot should we merge it or wait?

**Comment by mloskot** [comment]
I was going to test it against big-endian arch, but I haven't managed to dig for my QEMU environments. Sorry, no time really. If you want to go ahead and merege, no objections of course. Thanks.

File: include/boost/geometry/extensions/gis/io/wkb/detail/writer.hpp [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+				return true;
+			}
+		};
+
+	}} // namespace detail::wkb
+#endif // DOXYGEN_NO_IMPL
+
+}} // namespace boost::geometry
+
+#endif // BOOST_GEOMETRY_IO_WKB_DETAIL_WRITER_HPP
```
Comment (awulkiew) [inline include/boost/geometry/extensions/gis/io/wkb/detail/writer.hpp]: No newline at end of file

File: include/boost/geometry/extensions/gis/io/wkb/write_wkb.hpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	// 		{
+	// 			return false;
+	// 		}
+	// 
+	// 		return true;
+	// 	}
+
+}} // namespace boost::geometry
+
+#endif // BOOST_GEOMETRY_IO_WKB_WRITE_WKB_HPP
```
Comment (awulkiew) [inline include/boost/geometry/extensions/gis/io/wkb/write_wkb.hpp]: No newline at end of file.

File: extensions/test/gis/io/wkb/read_wkb.cpp [hunk 3]
```diff
@@ -22,87 +25,228 @@
...
+		// 
+		//     // POINT(1.234 5.678 15 79) - XYZM - Z and M compounds ignored
+		//     test_geometry_equals<point_type, true>(
+		//         "01010000C05839B4C876BEF33F83C0CAA145B616400000000000002E400000000000C05340",
+		//         "POINT (1.234 5.678)");
+		// 
+		//     // SRID=4326;POINT(1.234 5.678 15 79) - XYZM + SRID
+		//     test_geometry_equals<point_type, false>(
+		//         "01010000E0E61000005839B4C876BEF33F83C0CAA145B616400000000000002E400000000000C05340",
+		//         "POINT (1.234 5.678)");
```
Comment (awulkiew) [inline extensions/test/gis/io/wkb/read_wkb.cpp]: Why those test cases was commented out? Is there a regression?

File: include/boost/geometry/extensions/gis/io/wkb/detail/writer.hpp [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+				uint32_t num_points = boost::size(linestring);
+				value_writer<uint32_t>::write(num_points, iter, byte_order);
+
+				typedef typename point_type<L>::type point_type;
+
+				for(boost::range_iterator<const L>::type point_iter = boost::begin(linestring); point_iter != boost::end(linestring); ++point_iter)
+				{
+					// write point's x, y, z
+					writer_assigner<point_type, 0, dimension<point_type>::value>::run(*point_iter, iter, byte_order);
```
Comment (awulkiew) [inline include/boost/geometry/extensions/gis/io/wkb/detail/writer.hpp]: This and 133 lines are probably too long. In such cases we put the for-loop expressions or template parameters in separate lines, e.g.:

    for(boost::range_iterator<const L>::type point_iter = boost::begin(linestring);
        point_iter != boost::end(linestring);
        ++point_iter)
    {
        writer_assigner
            <
                point_type, 0, dimension<point_type>::value
            >::run(*point_iter, iter, byte_order);
    }

File: include/boost/geometry/extensions/gis/io/wkb/detail/writer.hpp [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+				value_writer<uint32_t>::write(num_exterior_ring_points, iter, byte_order);
+
+				for(
+					boost::range_iterator<const ring_type>::type point_iter = boost::begin(exterior_ring); 
+					point_iter != boost::end(exterior_ring);
+					++point_iter
+					)
+				{
+					// write point's x, y, z
+					writer_assigner<point_type, 0, dimension<point_type>::value>::run(*point_iter, iter, byte_order);
```
Comment (awulkiew) [inline include/boost/geometry/extensions/gis/io/wkb/detail/writer.hpp]: Here similar case, for-loop (empty line and big indentation) and template parameters of `writer_assigner`.

Btw there are some guidelines for developers, here: https://github.com/boostorg/geometry/wiki/Guidelines-for-Developers

**Comment by meastp** [comment]
Thank you, I will look at this and fix as soon as possible :)

**Comment by meastp** [comment]
@awulkiew : I have tried to fix all the violations of the guidelines I could find.

The other comment, about regressions in reading WKB (https://github.com/boostorg/geometry/pull/145#discussion-diff-25355264L25) :

The (removed) tests verified that one could e.g. read a PostGIS EWKB 3D point into a boost::geometry 2D point, where the WKB reader would implicitly ignore the last point. The reason I removed these tests was that this (surprising) behaviour will only ever work consistently with points – never with 3D vs 2D linestrings and polygons – and is therefore (imho) unwanted (and surprising) behaviour.

With this patch, 2D and 3D geometries are supported.

The patch has an unfinished implementation for (explicit) file formats (ogc wkb and ewkb in detail/ogc.hpp), but I am hoping this patch can be merged as is, and full file format support completed afterwards, in a separate pull request.

**Comment by meastp** [comment]
@awulkiew @mloskot (When the pull request is ready to merge, I (or you, as part of the merge) should probably it to squash the "fixup" commits)

**Comment by awulkiew** [comment]
Thanks the the fixes!
AFAIS there are still some "no newline at the end of file" issues.

**Comment by meastp** [comment]
@awulkiew : If I understand correctly, the issue is that there is a newline or carriage return at the end of some files. This contribution was made in Windows and Visual Studio, and I tried to remove the newline, with no luck. I also tried vi, unfortunately without luck.

I need help to be able to fix the newline at end of file issue(s)...

**Comment by awulkiew** [comment]
@meastp actually it's the opposite. There is no newline at the end of file but there should be. Some compilers will generate a warning if a file doesn't end with the newline.

**Comment by meastp** [comment]
@awulkiew : Thanks for explaining the issue. I have fixed this with the online github editor (which is why there is one commit per file). Are there any issues left? :)

**Comment by awulkiew** [comment]
@meastp There are tabs used for indentation. At Boost we don't use tabs. In particular in Boost.Geometry we use 4 spaces.

Would it be possible to squash the commits somehow? Most preferably grouping the changes made inside logical parts of the library. E.g. changes done in extensions headers and other for tests.
Or at least squash the commits containing all "guideline" fixes into one.

Also note, that in Boost.Geometry we add some tags in front of the commit message, e.g. in this case the tags could be `[extensions]` or `[extensions][wkb]` and `[extensions][test]`.

**Comment by meastp** [comment]
I will fix the remaining issues and do a rebase of the entire pull request to fix the commits. :)

**Comment by meastp** [comment]
@awulkiew I have fixed the remaining issues and squased the commits + rebase. It should be ready to merge :)

**Comment by mloskot** [comment]
:+1: Big thanks @meastp for the efforts and @awulkiew for taking the integration over

**Comment by awulkiew** [comment]
Thanks @meastp !

There is one small thing, i.e. instead of manually printing the debug info in  `test_geometry_equals()` `BOOST_CHECK_MESSAGE()` could be used. I'll change this after merging.

**Comment by barendgehrels** [comment]
Thanks a lot Mats! 
Very glad it is finished and merged now and will be part of Boost 1.59

**Comment by awulkiew** [comment]
It's in the extensions so it won't be released in 1.59 unless it's moved into the "official" part. Though AFAIU it's not ready.

Btw, there are some GCC compilation errors which I'm going to fix.

**Comment by barendgehrels** [comment]
Good to know. Why it is in the extensions? As far as I know there are no dependencies on other parts in extensions? What part is not ready (besides the fixing - thanks for that)?

**Comment by awulkiew** [comment]
Mats, please check out the fix and say if it's ok: https://github.com/boostorg/geometry/pull/282 e.g. if it doesn't interfere with your plans.

**Comment by meastp** [comment]
@awulkiew : Great! That's a better way to do it, and it does not interfere with the long(er) term goal of supporting multiple wkb formats (ogc, ewkb, ogc-sqlmm etc...). :)

**Pull request merged**